packages/multichain-account-service/src/providers/EvmAccountProvider.ts

```diff
@@ -76,10 +99,58 @@ export class EvmAccountProvider extends BaseBip44AccountProvider {
     return accountsArray;
   }
 
-  async discoverAndCreateAccounts(_: {
+  /**
+   * Discover and create accounts for the EVM provider.
+   *
+   * NOTE: This method should only be called on a newly created wallet.
+   *
+   * @param opts - The options for the discovery and creation of accounts.
+   * @param opts.entropySource - The entropy source to use for the discovery and creation of accounts.
+   * @param opts.groupIndex - The index of the group to create the accounts for.
+   * @returns The accounts for the EVM provider.
+   */
+  async discoverAndCreateAccounts(opts: {
     entropySource: EntropySourceId;
     groupIndex: number;
   }) {
-    return []; // TODO: Implement account discovery.
+    const provider = this.getEvmProvider();
+    const { entropySource, groupIndex } = opts;
+
+    const [address, didCreate] = await this.withKeyring<
+      EthKeyring,
+      [Hex, boolean]
+    >({ id: entropySource }, async ({ keyring }) => {
+      const existing = await keyring.getAccounts();
+      if (groupIndex < existing.length) {
+        return [existing[groupIndex], false];
+      }
+      const [added] = await keyring.addAccounts(1);
+      return [added, true];
+    });
+
+    const countHex = (await provider.request({
+      method: 'eth_getTransactionCount',
+      params: [address, 'latest'],
+    })) as Hex;
+    const count = parseInt(countHex, 16);
+
+    // We don't want to remove the account if it's the first one.
+    if (count === 0 && didCreate && groupIndex !== 0) {
+      await this.withKeyring<EthKeyring>(
+        { id: entropySource },
+        async ({ keyring }) => {
+          keyring.removeAccount?.(address);
+        },
+      );
+      return [];
+    }
```

Comment:
Should we wrap this in a `try finally`? So, if anything goes wrong with the provider request, we would still `removeAccount` the created account?